app/src/ui/dialog/header.tsx

```diff
@@ -4,8 +4,9 @@ import { Octicon, OcticonSymbol } from '../octicons'
 interface IDialogHeaderProps {
   /**
    * The dialog title text. Will be rendered top and center in a dialog.
+   * You can also pass a function that returns an element for custom styling.
    */
-  readonly title: string
+  readonly title: string | (() => JSX.Element)
```

Comment:
There's a gotcha here, because `dialog.tsx` has it's own title prop which is `string | undefined`.

Maybe we can add some examples of the sorts of styling changes we'd like to support here?

Reply:
@shiftkey the only example i have is the one in the PR description

Reply:
thanks @shiftkey @j-f1. I'm leaning towards the functional approach since the pattern of `renderSomething` is common in our components. personally, i would probably end up using a function to return the components even in the `React.ReactNode` implementation.

are there other benefits of taking a `React.ReactNode`? I'm honestly not familiar with that pattern.

Reply:
> personally, i would probably end up using a function to return the components even in the `React.ReactNode` implementation.

Can you provide a sample of how that would look? It'd be nice to assess the impact to callers, as I think there is a bit more work to uncover here about the API change.

> are there other benefits of taking a `React.ReactNode`? 

I think this is more relevant to React libraries, because it's a way to say you can handle "whatever React can render" rather than tying yourself to JSX/TSX explicitly.

Reply:
One of the benefits of taking `React.ReactNode` is that you can put JSX directly into the prop value without having to wrap it.

To remove reliance on string-like behavior, you could change code like that to use `JSON.stringify` since JSX is just objects.

Reply:
I think these are how the various cases would look like for the caller.

A `string`:
```tsx
<DialogHeader
  title='My cool title'
  dismissable={true}
/>
```

A `JSX.Element`:
```tsx

const renderTitle = () => (<span>My cool <strong>title</strong></span>)

<DialogHeader
  title={renderTitle}
  dismissable={true}
/>
```

A `React.ReactNode`:
```tsx

const titleNode = (<span>My cool <strong>title</strong></span>)

<DialogHeader
  title={titleNode}
  dismissable={true}
/>
```

I did a little research on the typings and I would prefer not to use react node since it can be a lot of things I don't want to be passed in here:

```ts
type ReactNode = ReactChild | ReactFragment | ReactPortal | string | number | boolean | null | undefined;
```

here's the type definition for `JSX.Element` for reference:
```ts
interface Element extends React.ReactElement<any> { }
```

Reply:
I'm 👍 on just sticking to `JSX.Element` for now, but I'm still not 💯% on the use of a function here, because it feels like it goes against the pattern of how we define our `render*` functions in the app. A random example is `SignIn`:

https://github.com/desktop/desktop/blob/a6138905a4ad0c52a607a9408e23f30b0754ef9a/app/src/ui/sign-in/sign-in.tsx#L291-L304

Our components can still use functions to generate the `JSX.Element`, but the app just executes them inline:

For example:

```tsx
private renderTitle(title: string) {
  return (
    <span>
      My cool <strong>{title}</strong>
    </span>
  )
}

public render() {
  ...
  return (
    <Dialog id="release-notes" onDismissed={this.props.onDismissed}>
      <DialogHeader title={this.renderTitle('branch')} dismissable={false}>
```

or it can be done inline if needed:

```tsx
public render() {
  ...

  const titleContents = 'branch'

  const title = (
    <span>
      My cool <strong>{titleContents}</strong>
    </span>
  )

  return (
    <Dialog id="release-notes" onDismissed={this.props.onDismissed}>
      <DialogHeader title={title} dismissable={false}>
```

Reply:
Your second example could be made a little shorter @outofambit:

```ts
<DialogHeader
  title={<span>My cool <strong>title</strong></span>}
  dismissable={true}
/>
```

Reply:
@shiftkey Are you thinking it would be better pass a straight up `JSX.Element` without the function wrapper instead? If so, I think I'm good with that now. :) (I think that's what you're basically saying too, @j-f1 ?)

Reply:
> Are you thinking it would be better pass a straight up JSX.Element without the function wrapper instead?

Yep, I'm 👍 with that approach (apologies for not including the signature change)